concordium-consensus/src/Concordium/KonsensusV1/Types.hs

```diff
@@ -0,0 +1,804 @@
+isAbsent Absent = True
+isAbsent (Present _) = False
+
+-- |The message that is signed by a finalizer to certify a block.
+data QuorumSignatureMessage = QuorumSignatureMessage
+    { -- |Hash of the genesis block.
+      qsmGenesis :: !BlockHash,
+      -- |Hash of the block being signed.
+      qsmBlock :: !BlockHash
+    }
+    deriving (Eq, Show)
+
+-- |Compute the byte representation of a 'QuorumSignatureMessage' that is actually signed.
+-- TODO: check that this cannot collide with other signed messages.
+quorumSignatureMessageBytes :: QuorumSignatureMessage -> BS.ByteString
+quorumSignatureMessageBytes QuorumSignatureMessage{..} = runPut $ do
+    putByteString "QUORUM."
+    put qsmGenesis
+    put qsmBlock
+
+-- |Signature by a finalizer on a 'QuorumSignatureMessage', or an aggregation of signatures on a
+-- common message.
+newtype QuorumSignature = QuorumSignature {theQuorumSignature :: Bls.Signature}
+    deriving (Eq, Ord, Show, Serialize, Semigroup, Monoid)
+
+-- |Sign a 'QuorumSignatureMessage' with a baker's private key.
+signQuorumSignatureMessage :: QuorumSignatureMessage -> BakerAggregationPrivateKey -> QuorumSignature
+signQuorumSignatureMessage msg privKey =
+    QuorumSignature $ Bls.sign (quorumSignatureMessageBytes msg) privKey
+
+-- |Check the signature on a 'QuorumSignatureMessage' that is signed by a single party.
+checkQuorumSignatureSingle ::
+    QuorumSignatureMessage -> BakerAggregationVerifyKey -> QuorumSignature -> Bool
+checkQuorumSignatureSingle msg pubKey =
+    Bls.verify (quorumSignatureMessageBytes msg) pubKey
+        . theQuorumSignature
+
+-- |Check the signature on a 'QuorumSignatureMessage' that is signed by multiple parties.
+checkQuorumSignature ::
+    QuorumSignatureMessage -> [BakerAggregationVerifyKey] -> QuorumSignature -> Bool
+checkQuorumSignature msg pubKeys =
+    Bls.verifyAggregate (quorumSignatureMessageBytes msg) pubKeys
+        . theQuorumSignature
+
+-- |Index of a finalizer in a finalization committee.
+newtype FinalizerIndex = FinalizerIndex {theFinalizerIndex :: Word32}
+    deriving (Eq, Ord, Show, Enum, Bounded, Serialize)
+
+-- |A set of 'FinalizerIndex'es.
+-- This is represented as a bit vector, where the bit @i@ is set iff the finalizer index @i@ is
+-- in the set.
+newtype FinalizerSet = FinalizerSet {theFinalizerSet :: Natural}
+    deriving (Eq)
+
+-- |The serialization of a 'FinalizerSet' consists of a length (Word32, big-endian), followed by
+-- that many bytes, the first of which (if any) must be non-zero. These bytes encode the bit-vector
+-- in big-endian. This enforces that the serialization of a finalizer set is unique.
+instance Serialize FinalizerSet where
+    put fs = do
+        let (byteCount, putBytes) = unroll 0 (return ()) (theFinalizerSet fs)
+        putWord32be byteCount
+        putBytes
+      where
+        unroll bc cont 0 = (bc, cont)
+        unroll bc cont n = unroll (bc + 1) (putWord8 (fromIntegral n) >> cont) (shiftR n 8)
```

Comment:
Please provide a type for this to make it easier to understand.

Reply:
And a sentence would help explain the "double reversal" thing going on to make this big endian.